joblib/numpy_pickle.py

```diff
@@ -6,32 +6,35 @@
 # Copyright (c) 2009 Gael Varoquaux
 # License: BSD Style, 3 clauses.
 
+import os
 import pickle
-import traceback
 import sys
-import os
-import zlib
+import traceback
 import warnings
-import struct
-import codecs
+import zlib
+from io import BytesIO
 
 from ._compat import _basestring
-
-from io import BytesIO
+from .externals import dill
 
 PY3 = sys.version_info[0] >= 3
+Pickler = dill.Pickler
+Unpickler = dill.Unpickler
 
-if PY3:
-    Unpickler = pickle._Unpickler
-    Pickler = pickle._Pickler
 
+def set_pickle(Pickler_, Unpickler_):
```

Comment:
I don't think this is going to work because when you call set_pickle, NumpyPickler and similar classes have already been defined and that will have no effect on them.

Also I am not sure you want to check the equality of the pickle with pickle and dill (re your test in test_numpy_pickle.py). The only thing you probably care about is backward-compatibility i.e. you need to make sure you can read pickle generated with 0.9.3 with 0.10. There is already a test like this (test_*across_python_version or something)